nfsv4/nfs4.c

```diff
@@ -1,189 +1,343 @@
                      const void *z,
                      int iAmt,
                      sqlite_int64 iOfst)
 {
-  return SQLITE_READONLY;
+    sqlfile f = (sqlfile)pFile;
+#ifdef TRACE
+    printf ("write %s offset:%lld bytes:%d\n", f->filename, iOfst, iAmt);
+#endif
+    translate_status(f->ad, writefile(f->f, (void *)z, iOfst, iAmt));
 }
 
 static int nfs4Truncate(sqlite3_file *pFile,
                         sqlite_int64 size)
 {
-  return SQLITE_READONLY;
+    sqlfile f = (sqlfile)pFile;
+#ifdef TRACE
+    printf ("truncate %s - WARNING NOT IMPLEMENTED\n", f->filename);
+#endif
+    return SQLITE_ERROR;
+
 }
 
 static int nfs4Sync(sqlite3_file *pFile, int flags)
-{
-  return SQLITE_READONLY;
+{ 
+   sqlfile f = (sqlfile)pFile;
+#ifdef TRACE
+   printf ("sync %s\n", f->filename);
+#endif
+    return SQLITE_OK;
 }
 
 static int nfs4FileSize(sqlite3_file *pFile, sqlite_int64 *pSize)
 {
-      file f = (file)pFile;
-      *pSize =  file_size(f->s, f->filename);
-      return SQLITE_OK;
+    sqlfile f = (sqlfile)pFile;
+#ifdef TRACE
+    printf ("filesize %s\n", f->filename);
+#endif    
+
+    u64 size;
+    status s =file_size(f->f, &size);
+    *pSize = size;
+    return translate_status(f->ad, s);
 }
 
 
-static int nfs4Lock(sqlite3_file *pFile, int eLock){
+static int nfs4Lock(sqlite3_file *pFile, int eLock)
+{
+#ifdef TRACE
+    printf ("lock %s\n", ((sqlfile)pFile)->filename);
+#endif        
     return SQLITE_OK;
 }
 
 static int nfs4Unlock(sqlite3_file *pFile, int eLock)
 {
+#ifdef TRACE
+    printf ("unlock %s\n", ((sqlfile)pFile)->filename);
+#endif        
     return SQLITE_OK;
 }
 
 static int nfs4CheckReservedLock(sqlite3_file *pFile, int *pResOut)
 {
+#ifdef TRACE
+    printf ("check lock %s\n", ((sqlfile)pFile)->filename);
+#endif            
     *pResOut = 0;
     return SQLITE_OK;
 }
 
 static int nfs4FileControl(sqlite3_file *pFile, int op, void *pArg)
 {
-  file f = (file)pFile;
-  int rc = SQLITE_OK;
+    sqlfile f = (sqlfile)pFile;
+#ifdef TRACE
+    printf ("control %s %d\n", f->filename, op);
+#endif                
 
-  if (op == SQLITE_FCNTL_MMAP_SIZE) {
-      *(u64 *) pArg = 0;
-  }
-      
-  if( op==SQLITE_FCNTL_VFSNAME ){
-      *(char**)pArg = sqlite3_mprintf("nfs4(%s)", f->filename);
-      rc = SQLITE_OK;
-  }
-  return rc;
+    int rc = SQLITE_OK;
+    
+    if (op == SQLITE_FCNTL_MMAP_SIZE) {
+        *(u64 *) pArg = 0;
+    }
+    
+    if( op==SQLITE_FCNTL_VFSNAME ){
+        buffer z = filename(f->f);
+        *(char**)pArg = sqlite3_mprintf("nfs4(%s)", z->contents + z->start);
+        rc = SQLITE_OK;
+    }
+    
+    return rc;
 }
 
-static int nfs4SectorSize(sqlite3_file *pFile){
+static int nfs4SectorSize(sqlite3_file *pFile)
+{
+    // see if we can get sqlite to use larger pages
     return 512;
 }
 
 static int nfs4DeviceCharacteristics(sqlite3_file *pFile){
-    return SQLITE_IOCAP_IMMUTABLE;
+    //set SQLITE_IOCAP_IMMUTABLE for read only
+    return 0;
```

Comment:
How about `SQLITE_IOCAP_POWERSAFE_OVERWRITE`? This is what the unix implementation does and from the comments it sounds like it may impact performance.

Reply:
see note about the .tables problem. I read the comments about this flag and completely escaped me what it accomplishes. will read again.

Reply:
I added POWERSAFE and didn't really see any difference in the number of writes. its possible that some of them are more contiguous than they were before. 

unfortunately, I'd really like to be able to pass flags in the .open to toggle this kind of thing (for this and future knobs), but unfortunately the query parameters of the file url aren't passed through :( (seems that would be natural)...looking into it